lua/entities/magicbutton.lua

```diff
@@ -0,0 +1,463 @@
+            trace1 = nil
+        end
+
+        if (trace1 == nil or TRACE1_FAILURES > 10) then
+            TRACE1_FAILURES = 0
+            trace1 = self:FindCastBox(startpoint) --returns a valid hull sized trace in a rnadom direction
+            TRACE_COUNTER = TRACE_COUNTER + 1
+            trace2 = nil
+        end
+
+        if (trace1 == nil) then
+            ORIGIN_FAILURES = ORIGIN_FAILURES + 1
+            TOTAL_FAILURES = TOTAL_FAILURES + 1
+            continue
+        else
+            trace2 = self:FindCastFinal(trace1, TRACE1_FAILURES) --only returns a final result if it's considered valid
+        end
+
+        if (trace2 == nil) then
+            TRACE1_FAILURES = TRACE1_FAILURES + 1
+            TOTAL_FAILURES = TOTAL_FAILURES + 1
+            continue
+        end
+    end
+
+    if (trace1 and trace2) then return trace2, trace1 end --print("SUCCESS AFTER "..TOTAL_FAILURES.." ATTEMPTS!")
+    --print("SUPER BIG FAILURE! HOW IS IT POSSIBLE?")
+end
+
+function ENT:MoveToTraceResult(trace)
+    self:SetPos(trace.HitPos)
+    local ang = trace.HitNormal:Angle()
+    ang:RotateAroundAxis(ang:Right(), -90)
+    ang:RotateAroundAxis(ang:Up(), 180)
+    self:SetAngles(ang)
+    self.HasSpot = true
+end
+
+function ENT:Initialize()
+    if (SERVER) then
+        self.Entity:SetModel("models/pyroteknik/secretbutton.mdl")
+        local bmins, bmaxs = Vector(-4.8, -3.1, 0), Vector(4.8, 3.1, 2)
+        self:SetCollisionBounds(bmins, bmaxs)
+        self.Entity:PhysicsInitBox(bmins, bmaxs)
+        self.Entity:SetMoveType(MOVETYPE_NONE)
+        self.Entity:SetCollisionGroup(COLLISION_GROUP_DEBRIS_TRIGGER)
+        self:SetUseType(SIMPLE_USE)
+        self:SetColor(HSVToColor(math.Rand(0, 360), 1, 1))
+        local phys = self:GetPhysicsObject()
+
+        if (IsValid(phys)) then
+            phys:EnableMotion(false)
+        end
+
+        --self:FindHidingSpot()
+        if (not self.HasSpot) then
+            local trace = self:FindHidingSpot()
+            self:MoveToTraceResult(trace)
+        end
+    end
+
+    if (CLIENT) then
+        local effectdata = EffectData()
+        effectdata:SetOrigin(self:GetPos())
+        effectdata:SetEntity(self)
+        util.Effect("propspawn", effectdata)
+    end
+end
+
+function ENT:SpawnFunction(ply, tr, ClassName)
+    local ent = ents.Create(ClassName)
+    ent:SetPos(tr.HitPos)
+    ent:SetAngles(VectorRand():AngleEx(tr.HitNormal))
+    local trace = ply:GetEyeTrace()
+    ent:MoveToTraceResult(trace)
+    ent.HasSpot = true
+    ent:Spawn()
+    ent:Activate()
+
+    return ent
+end
+
+function ENT:Think()
+end
+
+function ENT:OnRemove()
+end
+
+function ENT:Draw()
+    local pos = self:GetPos() + self:GetUp() * 3
+    local c = self:GetColor()
+    local cvector = (Vector(c.r, c.g, c.b) / 255)
+    local lc = render.ComputeLighting(pos, self:GetUp())
+    local light1 = {}
+    light1.type = MATERIAL_LIGHT_POINT
+    light1.color = cvector * 15
```

Comment:
So there is like a light on the button? what do you think about animating the intensity over time with `sin(SysTime())` ?